api/api-internal-package-repositories-v1/src/main/java/com/thoughtworks/go/apiv1/internalpackagerepositories/InternalPackageRepositoriesControllerV1.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2020 ThoughtWorks, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.thoughtworks.go.apiv1.internalpackagerepositories;
+
+import com.thoughtworks.go.api.ApiController;
+import com.thoughtworks.go.api.ApiVersion;
+import com.thoughtworks.go.api.CrudController;
+import com.thoughtworks.go.api.base.OutputWriter;
+import com.thoughtworks.go.api.spring.ApiAuthenticationHelper;
+import com.thoughtworks.go.apiv1.internalpackagerepositories.representers.PackageRepositoriesRepresenter;
+import com.thoughtworks.go.config.exceptions.EntityType;
+import com.thoughtworks.go.config.exceptions.HttpException;
+import com.thoughtworks.go.domain.packagerepository.PackageRepositories;
+import com.thoughtworks.go.server.service.EntityHashingService;
+import com.thoughtworks.go.server.service.materials.PackageRepositoryService;
+import com.thoughtworks.go.spark.Routes;
+import com.thoughtworks.go.spark.spring.SparkSpringController;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+import spark.Request;
+import spark.Response;
+
+import java.io.IOException;
+import java.util.function.Consumer;
+
+import static spark.Spark.*;
+
+@Component
+public class InternalPackageRepositoriesControllerV1 extends ApiController implements SparkSpringController {
+
+    private final ApiAuthenticationHelper apiAuthenticationHelper;
+    private final PackageRepositoryService packageRepositoryService;
+
+    @Autowired
+    public InternalPackageRepositoriesControllerV1(ApiAuthenticationHelper apiAuthenticationHelper, PackageRepositoryService packageRepositoryService) {
+        super(ApiVersion.v1);
+        this.apiAuthenticationHelper = apiAuthenticationHelper;
+        this.packageRepositoryService = packageRepositoryService;
+    }
+
+    @Override
+    public String controllerBasePath() {
+        return Routes.PackageRepository.INTERNAL_BASE;
+    }
+
+    @Override
+    public void setupRoutes() {
+        path(controllerBasePath(), () -> {
+             before("", mimeType, this::setContentType);
+             before("/*", mimeType, this::setContentType);
+
+            before("", mimeType, this.apiAuthenticationHelper::checkAdminUserOrGroupAdminUserAnd403);
```

Comment:
@kritika-singh3 - can we use the publicly documented [Package Repositories](https://api.gocd.org/current/#package-repositories) and [Packages](https://api.gocd.org/current/#packages) API to fetch this information.

IMO, we should always prefer public APIs for rendering entity information on the SPAs, unless its absolutely required to build a newer internal API.

Reply:
Reason for going with an internal API:
  - The public API for Package Repositories does not contain all the info related to the packages (i.e. `auto_update` and `configurations`)
 - Yes, I make two API calls to get all package repositories and packages. But, the package name is not unique across the GoCD instance - it is unique only inside a particular package repository. Hence, merging the two list (IMO) got a bit of a overhead (checking with both repo id and package name).

Let me know if you thing otherwise.

Reply:
As per the API documentation,

> The public API for Package Repositories does not contain all the info related to the packages (i.e. auto_update and configurations)

The package repositories API will not contain detailed packages information (as it is a referenced entity). But, the public packages API does contain the `auto_update` and `configurations` information.

> Yes, I make two API calls to get all package repositories and packages. But, the package name is not unique across the GoCD instance - it is unique only inside a particular package repository. Hence, merging the two list (IMO) got a bit of a overhead (checking with both repo id and package name).

What is the overhead? I mean, it is the part of the validation that both repo-id and package-name can not be duplicated, we're not doing anything extra.

---

Nevertheless, If we ourselves can not use the entity API on the entity SPA, then the usability of such API is questionable :)

Reply:
Agree with you. Will make the change.